Yul control flow graph (OptimizedEVMCodeTransform Step 1)

ekpyron opened the pull request:
First step involved in https://github.com/ethereum/solidity/pull/11493.

If anyone has a better idea for unit-testing this, I'm open to suggestions :-).

Still missing tests for stuff like ``continue``, ``break``, ``leave``, reverting control flow and probably more stuff.

~~Hm - and I moved this from ``libyul/backends/evm`` to ``libyul`` thinking that it's generic, but it actually isn't - the whole business of function return labels makes it rather evm specific, so maybe it should be moved back.~~

~~Actually, this should also mention the entries of each block in the test expectations, so that's also still missing.~~

ekpyron committed:
Control flow graph for Yul.

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `b6d7a56df896969dca51c22f0588edfcd825b788`:
```
Error: Trailing whitespace found:
test/libyul/yulControlFlowGraph/function.yul:28://   
test/libyul/yulControlFlowGraph/function.yul:36://   
test/libyul/yulControlFlowGraph/function.yul:41://   
test/libyul/yulControlFlowGraph/function.yul:48://   

```
Please check that your changes are working as intended.

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `49f791b6c81c7916d4b60bf645d2c1c0023c2b6d`:
```
Error: Trailing whitespace found:
test/libyul/yulControlFlowGraph/function.yul:28://   
test/libyul/yulControlFlowGraph/function.yul:36://   
test/libyul/yulControlFlowGraph/function.yul:41://   
test/libyul/yulControlFlowGraph/function.yul:48://   

```
Please check that your changes are working as intended.

chriseth commented on libyul/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+{
+	bool operator==(JunkSlot const&) const { return true; }
+	bool operator<(JunkSlot const&) const { return false; }
+	static constexpr bool canBeFreelyGenerated = true;
```
Is it possible to replace this by instead inheriting from a simple struct?

ekpyron replied:
Is that better?

ekpyron replied:
Or not sure I understand. I could have all slot types inherit from either ``CanBeFreelyGenerated`` or ``CanNotBeFreelyGenerated`` and move the ``static constexpr bool`` in there.
Or even from a ``template<bool v> struct CanBeFreelyGeneratedOrNot { static constexpr bool value = v; };``...
But what's the gain in that?

ekpyron replied:
There's no danger of a typo, if that's the concern. The free ``canBeFreelyGenerated`` function below will fail to compile, if that static member is missing in any of the variant types.

ekpyron replied:
I could also remove the members here and make the visit in the free function below specific for each type.

ekpyron commented:
I'm marking this "ready for review", even though it still needs more tests as mentioned in the description.
It's the first two commits of https://github.com/ethereum/solidity/pull/11493/.

There is the option, though, to change the way this works by having it rely on the disambiguator, for-loop-condition-into-body and expression splitter. https://github.com/ethereum/solidity/tree/newCodeGeneratorWithExpressionSplitter is a sketch of what that would look like. But I'm not sure if it's the right way to go. In its current state it generates less-optimal code than https://github.com/ethereum/solidity/pull/11493. Maybe that could be fixed, but that would probably involve to first take care of https://github.com/ethereum/solidity/pull/11493#discussion_r651808556 and https://github.com/ethereum/solidity/pull/11493#discussion_r651807831.

Generally, I think not going down that road and keeping it at the version in this PR might be better after all.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+
+void ControlFlowGraphBuilder::operator()(Block const& _block)
+{
+	ScopedSaveAndRestore saveScope(m_scope, m_info.scopes.at(&_block).get());
```
Maybe some cases should call `makeBlock()` here?

Or we should document that `BlockFlattner` (and therefore also disambiguator) should be run beforehand. 

```yul
{
    {
        let x := 0
    }
    let x := 0
}
// ----
// Block 0:
//   Entries: None
//   Assignment(x): [ 0x00 ] => [ x ]
//   Assignment(x): [ 0x00 ] => [ x ]
//   MainExit
```

Produces the same output as

```
{
    let x := 0
    x := 0
}
```

ekpyron replied:
That's what the whole scoping stuff does. It tracks scopes and resolves variables to ``Scope::Variable`` and functions to ``Scope::Function``, which are disambiguated.
So the graph itself is fine.
Not sure how to reproduce this in the tests, though. Unfortunately, the ``Scope::`` things work by means of every ``Scope::Variable`` and every ``Scope::Function`` living at separate memory locations and never being moved - so they're identified and compared by pointers, which are non-deterministic.
I'd change it, but it's rather deeply tied in the old code generator and elsewhere and rather critical to be done correctly...

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+	yulAssert(m_scope, "");
+	yulAssert(m_scope->identifiers.count(_function.name), "");
+	Scope::Function& function = std::get<Scope::Function>(m_scope->identifiers.at(_function.name));
+	m_graph.functions.emplace_back(&function);
```
Should function hoister be run beforehand? Currently, all nested function would go into global scope with this.

ekpyron replied:
See above. The ``Scope::Function`` things are disambiguated.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+	(*this)(_loop.pre);
+
+	std::optional<bool> constantCondition;
+	if (auto const* literalCondition = get_if<yul::Literal>(_loop.condition.get()))
```
Can we run loop condition into body so that this will always be true?

ekpyron replied:
I thought about it, but it's not that easy. You'll end up with more jumps in the graph, so to get the same results you'd then need to compress them later... which gets messy with simultanously tracking backwards jumps. Could probably be done, but not sure it's worth it.
In general the question is if we want to require to pre-transform the AST or run code generation on arbitrary ASTs. If the benefits of pre-transforming are minor, I'd say it's nicer to be able to run on arbitrary ASTs.
I did write a sketch of a version that goes further and even runs the expression splitter to avoid needing the ``TemporarySlot`` things altogether (it's https://github.com/ethereum/solidity/tree/newCodeGeneratorWithExpressionSplitter, but not sure how up to date that branch is) - but in the end wasn't convinced that it's really easier and beneficial (in fact the way the new code generator is currently written it does rely on, resp. benefit from the fact that statements are merged, if merging is compatible with the execution order).

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+			jump(loopBody, true);
+		}
+		else
+			jump(afterLoop);
```
I wonder if we can just skip this after applying loop condition into body.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+	(*this)(_loop.pre);
+
+	std::optional<bool> constantCondition;
+	if (auto const* literalCondition = get_if<yul::Literal>(_loop.condition.get()))
+	{
+		if (valueOfLiteral(*literalCondition) == 0)
+			constantCondition = false;
+		else
+			constantCondition = true;
+	}
```
```suggestion
	if (auto const* literalCondition = get_if<yul::Literal>(_loop.condition.get()))
		constantCondition = valueOfLiteral(*literalCondition) != 0;
```

ekpyron replied:
Lol :-D.

leonardoalt commented on test/libyul/yulControlFlowGraph/for.yul:
```diff
@@ -0,0 +1,31 @@
+//   Jump: 1
+// Block 1:
+//   Entries: 0, 2
+//   sload: [ 0x03 ] => [ TMP[sload, 0] ]
```
What's the 0 here?

ekpyron replied:
"Temporaries" are always return values of function calls - since functions can return multiple values, they are indexed, i.e. this is return value 0, i.e. the first return value, of a call to ``sload``.

leonardoalt commented on test/libyul/yulControlFlowGraph/function.yul:
```diff
@@ -0,0 +1,54 @@
+    h(y)
+}
+// ----
+// Block 0:
```
Are the `RET[f]` the function calls?

ekpyron replied:
Probably should have documented these somewhere :-). ``RET[f]`` is the return label of a specific call to ``f`` (it's a bit unfortunate that it's not easy to refer to some specific call more precisely than just using the function name).
For the record (it won't occur in this PR, but in later ones): There is also ``RET`` (without index), which refers to the return label of a function *within* the body of the function (i.e. not associated with a specific call and without a statically known value).

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+	static constexpr bool canBeFreelyGenerated = true;
+};
+using StackSlot = std::variant<FunctionCallReturnLabelSlot, FunctionReturnLabelSlot, VariableSlot, LiteralSlot, TemporarySlot, JunkSlot>;
+/// The stack top is usually the last element of the vector.
```
Why usually?

ekpyron replied:
Well, I can (and do) use this type to store unordered sets of stack elements at types as well (I can't use always a ``std::set``, since the comparison operators are non-deterministic and it's highly non-trivial to change them to *be* deterministic). Also I don't know if everybody who will ever use this thing will read the comment to stick to it, so I formulated it cautiously :-).

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+/// @returns true if @a _slot can be generated on the stack at any time.
+inline bool canBeFreelyGenerated(StackSlot const& _slot)
+{
+	return std::visit([](auto const& _typedSlot) { return std::decay_t<decltype(_typedSlot)>::canBeFreelyGenerated; }, _slot);
```
Do you really need `decay` here?

hrkrshnn replied:
I think for the `auto const&` to `auto` conversion.

ekpyron replied:
Pretty much, yeah. There's no reason to copy the slots for this, but I can't access the members in the type in a reference type, so I decay it. Would be a nice use case for C++20 templated lambda arguments.

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+	return std::visit([](auto const& _typedSlot) { return std::decay_t<decltype(_typedSlot)>::canBeFreelyGenerated; }, _slot);
+}
+
+/// Control flow graph consisting of ``DFG::BasicBlock``s connected by control flow.
```
```suggestion
/// Control flow graph consisting of ``CFG::BasicBlock``s connected by control flow.
```
?

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+	struct BuiltinCall
+	{
+		std::shared_ptr<DebugData const> debugData;
+		std::reference_wrapper<BuiltinFunction const> builtin;
```
Can't `builtin` be extracted from `functionCall`?

ekpyron replied:
One *could* query a dialect again for the builtin corresponding to ``functionCall.functionName.name`` later on, yeah... but there's little reason not to just store it here, is there? A lot of the information extracted into this graph (like the ``Scope::`` things) can be regenerated on the fly, but collecting all information required later for convenient access is among the main points of this whole thing :-).

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+		std::shared_ptr<DebugData const> debugData;
+		std::reference_wrapper<BuiltinFunction const> builtin;
+		std::reference_wrapper<yul::FunctionCall const> functionCall;
+		/// Number of proper arguments with a position on the stack, excluding literal arguments.
```
Why does it exclude literal arguments?

hrkrshnn replied:
I think it's just the case about things like `loadimmutable`, `setimmutable`, `dataoffset`, `datasize`, `memoryguard`, `linkersymbol`, and `verbatim`. Might actually be worth adding that as a comment.

ekpyron replied:
Added a bit more comment.

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+	struct FunctionCall
+	{
+		std::shared_ptr<DebugData const> debugData;
+		std::reference_wrapper<Scope::Function const> function;
```
Same here, can't `function` be extracted from `functionCall`?

ekpyron replied:
It can't really. Well it can, but it's highly non-trivial. To do so you'd need to keep track of scopes and resolve the function calls each time. Note that this isn't the ``FunctionDefinition``, but the disambiguated special ``Scope::Function`` thing the AsmAnalyzer produces. Doing this is here, s.t. I don't need to do it later is also one of the main points of the control flow graph (the subsequent steps are complex enough, so best to do as much as possible up front, s.t. the information is just there already later on).

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+		{
+			BasicBlock* target = nullptr;
+			/// The only backwards jumps are jumps from loop post to loop condition.
+			bool backwards = false;
```
Why do you need that information?

ekpyron replied:
Short answer: see https://github.com/ethereum/solidity/pull/11493 ;-).
Longer answer: the stack layout generator benefits from this information. The fact that Yul only contains loops and not arbitrary jumps is quite a good property for generating layouts - I can propagate the stack layouts from each exit upwards and further up to calling blocks through the entire graph ignoring backwards edges (which makes the graph acyclic) and I will still get the "correct" layouts at all for loop condition nodes - only the loop body is then not aware that it needs to keep the variables around that the condition needs, which the stack layout generator then "fixes" in a second pass.

If you have a more general solution that can generate stack layouts for arbitrarily cyclic control flow graphs that doesn't have higher runtime complexity, let me know, then I'll drop the information here ;-).

leonardoalt replied:
Nop, makes sense to keep it.

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+
+	/// Container for blocks for explicit ownership.
+	std::list<BasicBlock> blocks;
+	/// Container for creates variables for explicit ownership.
```
creates variables?

ekpyron replied:
create``d``. Actually ``generated`` might even be better.

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+	std::list<BasicBlock> blocks;
+	/// Container for creates variables for explicit ownership.
+	std::list<Scope::Variable> ghostVariables;
+	/// Container for creates calls for explicit ownership.
```
creates calls?

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+	/// Container for creates variables for explicit ownership.
+	std::list<Scope::Variable> ghostVariables;
+	/// Container for creates calls for explicit ownership.
+	std::list<yul::FunctionCall> ghostCalls;
```
What are ghost calls and variables?

ekpyron replied:
I didn't document that anywhere :-)? Doesn't look like I did...
I need ghost calls and ghost variables to transform switches to simpler control flow.
Basically
```
switch value
case 0 { ... }
case 1 { ... }
case 2 { ... }
default { ... }
```
is transformed to
```
let ghostVar := value
if eq(ghostVar, 0) // the "eq" calls being the ghost call
{
  ...
}
else if eq(ghostVar, 1) // well, I can't really write "else" in Yul, but conceptually that's what ends up in the CFG
{
 ....
}
else if eq(ghostVar, 2)
{
 ....
}
else
{
  ...
}
```
So here I need to generate one variable and three calls to ``eq`` for that.
The variable itself is particularly important if the switch value is a function call (which may have side-effects, so I may not be able to just duplicate it).

leonardoalt commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,193 @@
+}
+
+/// Control flow graph consisting of ``DFG::BasicBlock``s connected by control flow.
+struct CFG
```
I'm wondering if this could be extracted out of the EVM backend, that is, in case we want to have a Yul-only CFG in the future.

ekpyron replied:
I actually did that for a moment - but it's intertwined with the stack representation which in turn makes stuff like function return labels explicit, respectively it fixes the EVM calling convention, which does make it rather EVM specific, so I moved it back...

leonardoalt replied:
Ok!

leonardoalt commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/view/single.hpp>
+#include <range/v3/view/take_last.hpp>
+#include <range/v3/view/transform.hpp>
```
Maybe just include the whole thing? :p

ekpyron replied:
The whole thing is huge :-D. I do that at times to write small test snippets and even for those it noticably sucks for compile time :-D.

leonardoalt commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+			[](CFG::BasicBlock::Terminated&) {},
+			[](CFG::BasicBlock::MainExit&) {}
+		}, _node->exit);
+	});
```
Nice!

leonardoalt commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+	AsmAnalysisInfo& _analysisInfo,
+	Dialect const& _dialect
+):
+m_graph(_graph),
+m_info(_analysisInfo),
+m_dialect(_dialect)
```
```suggestion
    m_graph(_graph),
    m_info(_analysisInfo),
    m_dialect(_dialect)
```

ekpyron replied:
Yeah? I've at least seen it without being indented elsewhere, if I'm not just imagining things.

leonardoalt replied:
Maybe both are accepted then, idk.

ekpyron replied:
Changed it anyways.

leonardoalt commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+			})) |
+			ranges::views::values |
+			ranges::views::transform(std::ref(*this)) |
+			ranges::to<Stack>,
```
lol

leonardoalt replied:
I don't really understand what this sequence does. Can you maybe add a short comment?

hrkrshnn replied:
Does this look more understandable?

```cpp
vector<stackslot> s;
for (auto [idx, arg]: _call.arguments | enumerate | reverse)
    if (!builtin->literalArgument(idx).has_value())
        s.emplace_back(std::visit(*this, arg));
```

Basically just visit the arguments in reverse. Except for arguments that are not literals? For example, `loadimmutable`.

leonardoalt replied:
Ah ok, yea. But I think we can keep it the current way

chriseth replied:
I'm actually not sure if we should not replace it :)
The `transform(std::ref(*this))` call is very hidden.
Is it actually clear that these calls are done in order?

ekpyron replied:
Sure, the calls are done in order - there's no magic happening here, it's still basically range based loops, I'm just avoiding unnecessary variables :-). Also for the purpose of generating the graph the order of calls does not even matter. And not sure in what way ``transform(std::ref(*this))`` is "hidden" - it's an explicit transformation on the range spelled out right there :-).
That being said, I'll just rewrite it :-).

leonardoalt commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+				yulAssert(assignedVariables.size() == 1, "");
+				return {(*this)(_identifierOrLiteral)};
+			}
+		}, *_assignment.value),
```
This input code block is the same as in `operator()(vardecl)`. Maybe extract somehow?

leonardoalt commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+Scope::Variable const& ControlFlowGraphBuilder::lookupVariable(YulString _name) const
+{
+	yulAssert(m_scope, "");
+	Scope::Variable *var = nullptr;
```
```suggestion
	Scope::Variable* var = nullptr;
```

leonardoalt commented on test/libyul/ControlFlowGraphTest.cpp:
```diff
@@ -0,0 +1,207 @@
+	}
+	size_t getBlockId(CFG::BasicBlock const& _block)
+	{
+		if (size_t *id = util::valueOrNullptr(m_blockIds, &_block))
```
```suggestion
		if (size_t* id = util::valueOrNullptr(m_blockIds, &_block))
```

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+			// operation
+			CFG::BuiltinCall{_call.debugData, *builtin, _call}
+		});
+		std::get<CFG::BuiltinCall>(operation.operation).arguments = operation.input.size();
```
It might be more readable, if input was a temporary variable, and the size was also passed alongside `CFG::BuiltinCall` constructor in line 147.

hrkrshnn commented on test/libyul/ControlFlowGraphTest.cpp:
```diff
@@ -0,0 +1,207 @@
+		[](VariableSlot const& _var) { return _var.variable.get().name.str(); },
+		[](LiteralSlot const& _lit) { return util::toCompactHexWithPrefix(_lit.value); },
+		[](TemporarySlot const& _tmp) -> std::string { return "TMP[" + _tmp.call.get().functionName.name.str() + ", " + std::to_string(_tmp.index) + "]"; },
+		[](JunkSlot const&) -> std::string { return "JUNK"; }
```
This is never supposed to happen from the generator, right?

ekpyron replied:
No, it will only happen with the ``StackLayoutGenerator`` when merging control flow later on.
Example:
```
let x := 42
// Stack layout: [x]
if calldataload(0) {
  mstore(x, 0) // last use of x
}
// Since the "false"-branch directly jumps here, I do have ``x`` on stack. but it's never used again, so I the "entry"-layout
// of the post-if-block is ``[ JUNK ]``. This gives the exit of the "true"-branch the freedom to produce any value
// on the stack before jumping here instead of requiring it to keep a copy of ``x`` that's not needed.
```

chriseth commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+
+	// Determine which blocks are reachable from the entry.
+	util::BreadthFirstSearch<CFG::BasicBlock*> reachabilityCheck{{result->entry}};
+	ranges::actions::push_back(
+		reachabilityCheck.verticesToTraverse,
+		result->functionInfo | ranges::views::values | ranges::views::transform(
+			[](auto&& _function) { return _function.entry; }
+		)
+	);
```
Wouldn't a range-based loop be simpler and easier to understand?

chriseth commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+{
+	yulAssert(m_scope, "");
+	Scope::Variable *var = nullptr;
+	if (m_scope->lookup(_name, util::GenericVisitor{
```
Where is this function defined? What is the visitor used for? How does it cope with shadowing - does it stop at the first match?

ekpyron replied:
What function? ``m_scope->lookup``? That's the same old scoping mechanism used in old code generation generated during analysis. It's defined in ``libyul/Scope.h``. The visitor is only called once and only if the function returns ``true``. Not saying that this is designed particularly nicely, but at least for now I thought it best to just reuse the existing mechanism without changing it.

ekpyron replied:
Or put differently: compare with ``CodeTransform::operator()(Identifier const& _identifier)`` in the old ``EVMCodeTransform.cpp``, it's basically the same and in general it makes sense to share the mechanism.
If we want to change these, we should do so separately in both places.

chriseth commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,484 @@
+	}
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(FunctionCall const& _call)
```
Maybe better to group all `operator()` functions together.

cameel commented on test/tools/CMakeLists.txt:
```diff
@@ -41,4 +42,5 @@ add_executable(isoltest
 )
-target_link_libraries(isoltest PRIVATE evmc libsolc solidity yulInterpreter evmasm Boost::boost Boost::program_options Boost::unit_test_framework)
+target_compile_definitions(isoltest PRIVATE ISOLTEST)
+target_link_libraries(isoltest PRIVATE evmc libsolc solidity yulInterpreter evmasm Boost::boost Boost::program_options Boost::unit_test_framework Threads::Threads)
```
Why is `Threads::Threads` needed? I do not see it used anywhere (though I have not reviewed it all so maybe I just have not seen it yet).

ekpyron replied:
It's an implicit dependency of ``boost::process``, in particular ``child.detach();`` in ``StackLayoutGeneratorTest.cpp``

ekpyron replied:
Only breaks the static build - for all others I guess ``Boost::boost`` already pulls it in, but fails to do so for static builds. It's the main reason why I added the ``-DISOLTEST`` define, s.t. this is only really required for the interactive runs.

cameel replied:
Oh, thanks for mentioning `child.detach()`. I just did not get to that part yet. So basically you need it because you run graphviz as a child process.

It opens a window like you have shown on the call and keeps isoltest waiting in the background until you close it? Or does it render to file?

ekpyron replied:
It passes the graph as ``stdin`` to whatever you specify as environment variables, which can do whatever you want with it.
The intended purpose is to have it open a window with the rendered graph, though, as in setting the environment variables to ``dot -Tx11:cairo`` or ``xdot -`` or something like that.

cameel commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,199 @@
+{
+	std::reference_wrapper<yul::FunctionCall const> call;
+	bool operator==(FunctionCallReturnLabelSlot const& _rhs) const { return &call.get() == &_rhs.call.get(); }
+	bool operator<(FunctionCallReturnLabelSlot const& _rhs) const { return &call.get() < &_rhs.call.get(); }
```
Comparing pointers may result in the order changing between compiler runs. Can the order affect generated EVM code?

ekpyron replied:
The order is indeed nondeterministic, but it's only needed them to construct ``set``s and ``map``s of stack slots, none of which are traversed, so the order won't affect the result. It would of course be nicer to have a deterministic order, but it's nontrivial to construct one.

cameel replied:
ok. Seems fine then.

cameel commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,199 @@
+	static constexpr bool canBeFreelyGenerated = true;
+};
+/// A slot containing the idx-th return value of a previous call.
+struct TemporarySlot
```
Why not `ReturnSlot`?

ekpyron replied:
``ReturnSlot`` to me rather sounds like something containing return values within a function...
These slots are what I get on stack immediately after a function call, which will either be used as input for another call or immediately given a proper variable name, so in that sense I'd say "temporary" matches pretty well - they're always short-lived.

ekpyron replied:
But I'd also be fine with renaming them.

cameel replied:
Well, ok, I thought that these slots and the slots used inside function call would basically be the same thing. If they're different things in the graph then yeah, maybe `ReturnSlot` would make things ambiguous.

cameel commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,199 @@
+	/// Maintains a list of entry blocks and a typed exit.
+	struct BasicBlock
+	{
+		std::vector<BasicBlock*> entries;
+		std::vector<Operation> operations;
```
I'd keep all members grouped together. It's easy to miss some of them if they're interspersed with types. For a moment here I was wondering how it makes sense for `exit` to be the only member in this struct :)

ekpyron replied:
Fair enough :-). Changed it.

cameel reviewed (commented):
I did not manage to review it all yet. For now just a few comments, mostly questions.

cameel commented on libyul/Scope.h:
```diff
@@ -37,11 +37,12 @@ struct Scope
 	using YulType = YulString;
 
-	struct Variable { YulType type; };
+	struct Variable { YulType type; YulString name; };
```
```suggestion
	struct Variable {
		YulType type;
		YulString name;
	};
```

cameel commented on test/libyul/ControlFlowGraphTest.cpp:
```diff
@@ -0,0 +1,270 @@
+
+#ifdef ISOLTEST
+	char* graphDisplayer = nullptr;
+	if (result == TestResult::Failure)
+		graphDisplayer = getenv("ISOLTEST_DISPLAY_GRAPHS_FAILURE");
+	else if (result == TestResult::Success)
+		graphDisplayer = getenv("ISOLTEST_DISPLAY_GRAPHS_SUCCESS");
```
I know it's already long, but having `COMMAND` in the name makes it clearer what the expected value of the variable is (with the current name I would not really expect it to be the name of an executable):
```suggestion
	if (result == TestResult::Failure)
		graphDisplayer = getenv("ISOLTEST_FAILURE_DISPLAY_GRAPHS_COMMAND");
	else if (result == TestResult::Success)
		graphDisplayer = getenv("ISOLTEST_SUCCESS_DISPLAY_GRAPHS_COMMAND");
```
- We should document these variables somewhere.
- Why different commands for success and failure?
- The program you use for this is `dot`, right? Maybe it should be the default value?

ekpyron replied:
Displaying something in the success and failure case have rather orthogonal use cases.
If you work on this stuff and change something, you'll want to have it show you the failure cases and just pass through on the success cases.
If you review a PR and want to verify test cases, you'll want to have it stop at each case and show you the graph.
I don't think there's a reasonable default value for this.
You can use ``dot -Tx11:cairo``, but that's already platform specific and sucks interface wise. You can use ``xdot``, if you have it installed and it's available on your platform, but you know neither.

I don't think this will actually be done regularly later on, so I don't think it's worth over-engineering.

ekpyron replied:
Renamed them in any case.

ekpyron replied:
Also added some comments.

cameel commented on test/libyul/ControlFlowGraphTest.cpp:
```diff
@@ -0,0 +1,270 @@
+		pipe << output.str();
+		pipe.flush();
+		pipe.pipe().close();
+		if (result == TestResult::Success)
+			child.wait();
+		else
+			child.detach();
```
If you detach on failure, won't this just result in isoltest leaving spawned child processes running in the background?

ekpyron replied:
Yes. In the failure case, ``isoltest`` will display the test case and a ``diff`` and stop on its own, but only after you return from here, so this is intentional.
In the success case it won't stop on its own and not output anything, so I output the source manually here and have the child process block.

cameel commented on libyul/backends/evm/ControlFlowGraphBuilder.h:
```diff
@@ -0,0 +1,81 @@
+		AsmAnalysisInfo& _analysisInfo,
+		Dialect const& _dialect
+	);
+	CFG::Operation& visitFunctionCall(FunctionCall const&);
```
I think the result could be `const` here. In all uses I've seen the result is either only read from or used to return a copy of the stack.
```suggestion
	CFG::Operation const& visitFunctionCall(FunctionCall const&);
```

cameel commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,467 @@
+	});
+
+	// Remove all entries from unreachable nodes from the graph.
+	for (auto* node: reachabilityCheck.visited)
```
It's just `BasicBlock`, not something more complex, right?
```suggestion
	for (CFG::BasicBlock* node: reachabilityCheck.visited)
```

ekpyron replied:
We should really have that team-wide discussion about `auto` we wanted to have :-D. But to cut this short, I'll just change it :-).

cameel replied:
I thought that so far the status quo was to use `auto` only if the type is particularly complex (iterator, nested template parameters, etc) or when it's already used inside the expression (e.g. in a cast).

In any case, a lot of the time I prefer to see the type rather than `auto` because otherwise I often have to either look it up myself or actually switch to the PR branch locally and open it in the editor where I can just inspect it (but it's a bit tedious to do this for every PR).

cameel commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,467 @@
+				return TemporarySlot{_call, _i};
+			}) | ranges::to<Stack>,
+			// operation
+			std::move(builtinCall)
```
```suggestion
			CFG::BuiltinCall{_call.debugData, *builtin, _call, inputs.size()};
```

cameel commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,467 @@
+			if (!builtin->literalArgument(idx).has_value())
+				inputs.emplace_back(std::visit(*this, arg));
+		CFG::BuiltinCall builtinCall{_call.debugData, *builtin, _call, inputs.size()};
+		return m_currentBlock->operations.emplace_back(CFG::Operation{
```
Since this is `emplace`, I think you should specify constructor arguments directly here. Otherwise it runs the copy constructor so you could just as well do `push_back()`.
```suggestion
		return m_currentBlock->operations.emplace_back(
```

ekpyron replied:
Can't unless I add a constructor to the thing. In contrast to using ``push_back`` this *won't* copy, but move the operation and the compiler will probably be clever enough to eliminate the move here, since it's an implicit move constructor anyways (it's allowed to).

cameel replied:
Oh, fine then.

cameel commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,467 @@
+Scope::Variable const& ControlFlowGraphBuilder::lookupVariable(YulString _name) const
+{
+	yulAssert(m_scope, "");
+	Scope::Variable* var = nullptr;
```
```suggestion
	Scope::Variable const* var = nullptr;
```

cameel commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,467 @@
+	}
+	else
+	{
+		Scope::Function* function = nullptr;
+		yulAssert(m_scope->lookup(_call.functionName.name, util::GenericVisitor{
+			[](Scope::Variable&) { yulAssert(false, "Expected function name."); },
+			[&](Scope::Function& _function) { function = &_function; }
+		}), "Function name not found.");
+		yulAssert(function, "");
```
How about putting it in `lookupFunction()` analogous to `lookupVariable()` that already exists? It looks unnecessarily complex for what it actually does and hiding that in a function would make this code look a bit simpler.

ekpyron replied:
Yeah, ideally we'd change this weird mechanism in the Scoping class, but I won't do that here - but extracting to a function sounds good!

hrkrshnn commented on libsolutil/cxx20.h:
```diff
@@ -49,4 +49,15 @@ erase_if(std::unordered_map<Key, T, Hash, KeyEqual, Alloc>& _c, Pred _pred)
+	auto it = std::remove_if(c.begin(), c.end(), pred);
+	auto r = std::distance(it, c.end());
+	c.erase(it, c.end());
+	return static_cast<typename std::vector<T, Alloc>::size_type>(r);
```
Is the static cast required here? The code at cppreference doesn't have it. Fine with keeping it though.

ekpyron replied:
It's needed, because we enable conversion warnings and treat them fatal - I wouldn't have put it in otherwise ;-).

hrkrshnn commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,199 @@
+/// The return label of a function while generating the code of the function body.
+struct FunctionReturnLabelSlot
+{
+	bool operator==(FunctionReturnLabelSlot const&) const { return true; }
```
Why are these comparisons needed? My not compare pointers to `functionDefinition`?

ekpyron replied:
You've seen https://github.com/ethereum/solidity/pull/11615, haven't you :-)? I'll end up equality comparing stack slots all the time :-).
(``StackSlot``, which is a ``std::variant``, will inherit comparisons following https://en.cppreference.com/w/cpp/utility/variant/operator_cmp, but only if all members have comparison operations)
If I wanted to compare pointers to some ``functionDefinition``, I'd actually need to store some such pointer first... and it's not really needed.
We could at some point even merge matching subgraphs of the control flow graph and have one such slot refer to the return label of multiple function. Or more realistically, we might do tail-call-optimization at some point. There's nothing really specific to a function definition about this slot - it's just the label the last jump into any function wanted to have the next function return return to.

cameel commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,472 @@
+		[&](auto const&) { yulAssert(false, ""); }
+	}, _exprStmt.expression);
+
+	// TODO: Ideally this would be done on the expression label and for all functions that always revert,
+	//       not only for builtins.
```
Maybe this should be an issue?

ekpyron replied:
I don't think too fine-grained issues make too much sense right now. It's too much of a hassle to change things like this before the entire series of PRs down to actual code generation is merged anyways - once that's done, I'll create a detailed description of everything that should still be improved afterwards.

cameel replied:
ok. fine.

cameel commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,472 @@
+		jump(afterSwitch);
+		m_currentBlock = elseBranch;
+	}
+	Case const& switchCase = _switch.cases.back();
```
I'd assert that the `switch` does not have zero cases.

ekpyron replied:
It's guaranteed by analysis, but another assert doesn't hurt, I'll add it.

cameel replied:
Even if it's guaranteed by something somewhere far away I usually prefer to have a local assert (if it's not too complex to add one) when my code depends on it :)

cameel commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,472 @@
+		(*this)(switchCase.body);
+	}
+	else
+		(*this)(switchCase.body);
```
```suggestion
	}
	(*this)(switchCase.body);
```

hrkrshnn commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,199 @@
+struct FunctionCallReturnLabelSlot
+{
+	std::reference_wrapper<yul::FunctionCall const> call;
+	bool operator==(FunctionCallReturnLabelSlot const& _rhs) const { return &call.get() == &_rhs.call.get(); }
```
Is there an easy way to tell what slots also needs the debugData?

ekpyron replied:
All slots for which there *is* a suitable choice of ``debugData`` to associate them with and that choice is not just the ``debugData`` of the single Yul AST node stored in it.
That being said I could also just have them everywhere, if you prefer...

cameel commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,472 @@
+		}) | ranges::to<vector>
+	}));
+	yulAssert(inserted, "");
+	CFG::FunctionInfo& info = it->second;
```
Having `m_info` of type `AnalysisInfo` and `info` of tyle `FunctionInfo` makes this part a bit confusing.
```suggestion
	CFG::FunctionInfo& functionInfo = it->second;
```

cameel reviewed (commented):
Here's what i got so far. Again, nothing serious, just small tweaks.

I still haven't managed to review tests but I'm out of time now. I'll get back to this tomorrow.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,199 @@
+	CFG(CFG const&) = delete;
+	CFG(CFG&&) = delete;
+	CFG& operator=(CFG const&) = delete;
+	CFG& operator=(CFG&&) = delete;
```
Shouldn't we define a destructor? http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-five

ekpyron replied:
Can do it.

ekpyron commented:
Addressed all comments again.
Also in case this comes up ;-): if we want to just have more test cases, I suggest to collect cases we want and add them later on, but not keep this PR pending because of it. The graph won't be used for anything until the follow-up PRs are merged, so we can extend test coverage in parallel.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,199 @@
+struct VariableSlot
+{
+	std::reference_wrapper<Scope::Variable const> variable;
+	std::shared_ptr<DebugData const> debugData{};
```
```suggestion
	std::shared_ptr<DebugData const> debugData;
```

Optional. Couple more below too.

ekpyron replied:
They're not optional, they are required to be able to omit them in aggregate initialization.

cameel commented:
> Also in case this comes up ;-): if we want to just have more test cases, I suggest to collect cases we want and add them later on, but not keep this PR pending because of it.

Haha, it will definitely come up :) I'm fine with adding a more complete coverage only in later PRs though. I saw you have already put a lot of effort into testing it.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.h:
```diff
@@ -0,0 +1,82 @@
+	void makeConditionalJump(StackSlot _condition, CFG::BasicBlock& _nonZero, CFG::BasicBlock& _zero);
+	void jump(CFG::BasicBlock& _target, bool _backwards = false);
+	CFG& m_graph;
+	AsmAnalysisInfo& m_info;
```
```suggestion
	AsmAnalysisInfo const& m_info;
```

hrkrshnn replied:
Not sure if other PRs create a problem for this.

ekpyron replied:
That's actually nice! I just assumed ``AsmAnalysisInfo`` to have ill-defined ``const``ness, since the old code transform stores a mutable reference to it, but apparently that's not the case, so this works and I'll change it.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,471 @@
+		cxx20::erase_if(node->entries, [&](CFG::BasicBlock* entry) -> bool {
+			return !reachabilityCheck.visited.count(entry);
+		});
+
```
Why do we need to do this pre-processing? I think the optimizer should cover this anyway, so maybe it's redundant?

ekpyron replied:
It doesn't :-).

ekpyron replied:
Also the reason is for this actually not optimization, but fixing the graph.
At a reverting builtin call, the graph will let the current block end in a "termination exit" and start a new block (it needs to to continue the visit). But none of the blocks after that termination exit will be reachable - however, they'll still contribute to the ``entries`` field of subsequent node. So in the end we need to reduce the entry nodes to nodes that are actually reachable to have a proper graph.

ekpyron replied:
An alternative would be to not generate the entry fields initially at all and only fill them in afterwards in a visit...

ekpyron commented:
On the other hand we really have to take care that we test the hell out of it before we actually enable the new code transform for anything. The new code transform will basically be able to - at least to some degree - self-assert its correctness, but only relative to this graph, so it has to be rock-solid. But yeah, to feasibly make progress on it, it's probably wise to spread it out a bit.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,471 @@
+			yulAssert(_expectedSlotCount == operation.output.size(), "");
+			return operation.output;
+		},
+		[&](auto const& _identifierOrLiteral) -> Stack{
```
```suggestion
		[&](auto const& _identifierOrLiteral) -> Stack {
```

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,471 @@
+{
+	vector<VariableSlot> assignedVariables = _assignment.variableNames | ranges::views::transform([&](Identifier const& _var) {
+		return VariableSlot{lookupVariable(_var.name), _var.debugData};
+	}) | ranges::to<vector<VariableSlot>>;
```
```suggestion
	}) | ranges::to<vector>;
```

Just for consistency with line 133. Fine otherwise too.

ekpyron replied:
Deduction for those doesn't always work on all platforms... but I can change the other one for consistency.

chriseth commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,200 @@
+	bool operator<(FunctionCallReturnLabelSlot const& _rhs) const { return &call.get() < &_rhs.call.get(); }
+	static constexpr bool canBeFreelyGenerated = true;
+};
+/// The return label of a function while generating the code of the function body.
```
```suggestion
/// The return label of a function while generating the code of the function body.
```
please clarify that this is not the target of the `leave` statement but the actual return target of the whole function passed in as a FunctionCallReturnLabelSlot

chriseth commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,200 @@
+	bool operator<(LiteralSlot const& _rhs) const { return value < _rhs.value; }
+	static constexpr bool canBeFreelyGenerated = true;
+};
+/// A slot containing the idx-th return value of a previous call.
```
```suggestion
/// A slot containing the index-th return value of a previous call.
```

chriseth commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,471 @@
+
+	reachabilityCheck.run([&](CFG::BasicBlock* _node, auto&& _addChild) {
+		visit(util::GenericVisitor{
+			[&](CFG::BasicBlock::Jump& _jump) {
```
```suggestion
			[&](CFG::BasicBlock::Jump const& _jump) {
```
?

chriseth commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,471 @@
+	if (_varDecl.value)
+		input = visitAssignmentRightHandSide(*_varDecl.value, declaredVariables.size());
+	else
+		input = ranges::views::iota(0u, _varDecl.variables.size()) | ranges::views::transform([&](size_t) {
+			return LiteralSlot{0, _varDecl.debugData};
+		}) | ranges::to<Stack>;
```
Is this the same as
```suggestion
		input = Stack(_varDecl.variables.size(), LiteralSlot{0, _varDecl.debugData});
```
?

ekpyron replied:
Haha, yeah, it is :-).

chriseth commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,471 @@
+		}) | ranges::to<Stack>;
+	m_currentBlock->operations.emplace_back(CFG::Operation{
+		std::move(input),
+		declaredVariables | ranges::to<Stack>,
```
Why don't you convert it to a `Stack` at the place of its declaration?

chriseth replied:
Oh because of the next line?

ekpyron replied:
It's also stored in ``CFG::Assignment`` further down as actual ``vector<VariableSlot>``. Strictly speaking it's redundant to also store it there, but it's convenient to have a list of assigned variables directly in ``CFG::Assignment`` there later on. But I can also remove it and later reconstruct it from the operation output instead.

chriseth commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,471 @@
+		[&](auto const&) { yulAssert(false, ""); }
+	}, _exprStmt.expression);
+
+	// TODO: Ideally this would be done on the expression label and for all functions that always revert,
```
We should have a side effect computer for that, but maybe it's still a "todo"?

ekpyron replied:
Yeah, not sure we have a computer for control flow side effects yet - and the idea was to improve it in a second step after this first version.

ekpyron commented:
Rebased.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.h:
```diff
@@ -0,0 +1,82 @@
+
+	Scope::Function const& lookupFunction(YulString _name) const;
+	Scope::Variable const& lookupVariable(YulString _name) const;
+	std::pair<CFG::BasicBlock*, CFG::BasicBlock*> makeConditionalJump(StackSlot _condition);
```
Missing documentation for pair. Mainly, which one is zero and which one is non-zero.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+	};
+	_nonZero.entries.emplace_back(m_currentBlock);
+	_zero.entries.emplace_back(m_currentBlock);
+	m_currentBlock = nullptr;
```
Why this assignment? It always seems to be reassigned afterwards. Do you simply want to enforce that it always gets reassigned?

ekpyron replied:
Yes, the idea is exactly to enforce that it always gets reassigned explicitly afterwards.

hrkrshnn reviewed (commented):
None of the tests check the scoping mechanism, i.e., variables and functions have unique names.

```
{
  a()
  c()
  function a() {
    b()
    function b() {}
  }
  function c() {
    b()
    function b() {}
  }
}
```

Graphviz seems to have a problem with this example. See [visualizer](https://dreampuf.github.io/GraphvizOnline/#%20%20%20%20digraph%20CFG%20%7B%0A%20%20%20%20nodesep%3D0.7%3B%0A%20%20%20%20node%5Bshape%3Dbox%5D%3B%0A%0A%20%20%20%20Entry%20%5Blabel%3D%22Entry%22%5D%3B%0A%20%20%20%20Entry%20-%3E%20Block0%3B%0A%20%20%20%20Block0%20%5Blabel%3D%22%5C%0A%20%20%20%20a%3A%20%5B%20RET%5Ba%5D%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20c%3A%20%5B%20RET%5Bc%5D%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block0Exit%20%5Blabel%3D%22MainExit%22%5D%3B%0A%20%20%20%20Block0%20-%3E%20Block0Exit%3B%0A%0A%20%20%20%20FunctionEntry_a%20%5Blabel%3D%22function%20a()%22%5D%3B%0A%20%20%20%20FunctionEntry_a%20-%3E%20Block1%3B%0A%20%20%20%20Block1%20%5Blabel%3D%22%5C%0A%20%20%20%20b%3A%20%5B%20RET%5Bb%5D%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block1Exit%20%5Blabel%3D%22FunctionReturn%5Ba%5D%22%5D%3B%0A%20%20%20%20Block1%20-%3E%20Block1Exit%3B%0A%0A%20%20%20%20FunctionEntry_b%20%5Blabel%3D%22function%20b()%22%5D%3B%0A%20%20%20%20FunctionEntry_b%20-%3E%20Block2%3B%0A%20%20%20%20Block2%20%5Blabel%3D%22%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block2Exit%20%5Blabel%3D%22FunctionReturn%5Bb%5D%22%5D%3B%0A%20%20%20%20Block2%20-%3E%20Block2Exit%3B%0A%0A%20%20%20%20FunctionEntry_c%20%5Blabel%3D%22function%20c()%22%5D%3B%0A%20%20%20%20FunctionEntry_c%20-%3E%20Block3%3B%0A%20%20%20%20Block3%20%5Blabel%3D%22%5C%0A%20%20%20%20b%3A%20%5B%20RET%5Bb%5D%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block3Exit%20%5Blabel%3D%22FunctionReturn%5Bc%5D%22%5D%3B%0A%20%20%20%20Block3%20-%3E%20Block3Exit%3B%0A%0A%20%20%20%20FunctionEntry_b%20%5Blabel%3D%22function%20b()%22%5D%3B%0A%20%20%20%20FunctionEntry_b%20-%3E%20Block4%3B%0A%20%20%20%20Block4%20%5Blabel%3D%22%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block4Exit%20%5Blabel%3D%22FunctionReturn%5Bb%5D%22%5D%3B%0A%20%20%20%20Block4%20-%3E%20Block4Exit%3B%0A%0A%20%20%20%20%7D%0A)

ekpyron commented:
> None of the tests check the scoping mechanism, i.e., variables and functions have unique names.
> 
> ```
> {
>   a()
>   c()
>   function a() {
>     b()
>     function b() {}
>   }
>   function c() {
>     b()
>     function b() {}
>   }
> }
> ```
> 
> Graphviz seems to have a problem with this example. See [visualizer](https://dreampuf.github.io/GraphvizOnline/#%20%20%20%20digraph%20CFG%20%7B%0A%20%20%20%20nodesep%3D0.7%3B%0A%20%20%20%20node%5Bshape%3Dbox%5D%3B%0A%0A%20%20%20%20Entry%20%5Blabel%3D%22Entry%22%5D%3B%0A%20%20%20%20Entry%20-%3E%20Block0%3B%0A%20%20%20%20Block0%20%5Blabel%3D%22%5C%0A%20%20%20%20a%3A%20%5B%20RET%5Ba%5D%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20c%3A%20%5B%20RET%5Bc%5D%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block0Exit%20%5Blabel%3D%22MainExit%22%5D%3B%0A%20%20%20%20Block0%20-%3E%20Block0Exit%3B%0A%0A%20%20%20%20FunctionEntry_a%20%5Blabel%3D%22function%20a()%22%5D%3B%0A%20%20%20%20FunctionEntry_a%20-%3E%20Block1%3B%0A%20%20%20%20Block1%20%5Blabel%3D%22%5C%0A%20%20%20%20b%3A%20%5B%20RET%5Bb%5D%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block1Exit%20%5Blabel%3D%22FunctionReturn%5Ba%5D%22%5D%3B%0A%20%20%20%20Block1%20-%3E%20Block1Exit%3B%0A%0A%20%20%20%20FunctionEntry_b%20%5Blabel%3D%22function%20b()%22%5D%3B%0A%20%20%20%20FunctionEntry_b%20-%3E%20Block2%3B%0A%20%20%20%20Block2%20%5Blabel%3D%22%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block2Exit%20%5Blabel%3D%22FunctionReturn%5Bb%5D%22%5D%3B%0A%20%20%20%20Block2%20-%3E%20Block2Exit%3B%0A%0A%20%20%20%20FunctionEntry_c%20%5Blabel%3D%22function%20c()%22%5D%3B%0A%20%20%20%20FunctionEntry_c%20-%3E%20Block3%3B%0A%20%20%20%20Block3%20%5Blabel%3D%22%5C%0A%20%20%20%20b%3A%20%5B%20RET%5Bb%5D%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block3Exit%20%5Blabel%3D%22FunctionReturn%5Bc%5D%22%5D%3B%0A%20%20%20%20Block3%20-%3E%20Block3Exit%3B%0A%0A%20%20%20%20FunctionEntry_b%20%5Blabel%3D%22function%20b()%22%5D%3B%0A%20%20%20%20FunctionEntry_b%20-%3E%20Block4%3B%0A%20%20%20%20Block4%20%5Blabel%3D%22%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block4Exit%20%5Blabel%3D%22FunctionReturn%5Bb%5D%22%5D%3B%0A%20%20%20%20Block4%20-%3E%20Block4Exit%3B%0A%0A%20%20%20%20%7D%0A)

[For the record basically what I just said on the matrix channel] Well, they *don't* have unique names, the preexisting (and thus here not explicitly tested) scoping mechanism just maps them to unique pointers, but those I can't really display in the tests, so I settle for the names instead, even though they are ambiguous, so it's not unexpected that this won't show up nicely in the tests. If this is bothersome, I can just disambiguate the AST after all, that's easier than trying to artifically choose determinstic unique names for the scope pointers.

ekpyron commented:
@hrkrshnn I changed my mind and just pushed a small change to the tests, at least adding the body block ID of a function to the graph node name, which makes them work nicely with graphviz, that still won't actually show different names, but collapse several ``Scope::*`` pointer thingys for identifiers with the same name, but at least the graph will still show nicely.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+		CFG::BasicBlock& caseBranch = m_graph.makeBlock();
+		makeConditionalJump(makeValueCompare(*switchCase.value), caseBranch, afterSwitch);
+		m_currentBlock = &caseBranch;
+	}
```
For 

```
{
    sstore(0, 0)
    switch sload(0)
    case 0 {
        sstore(0x01, 0x0101)
    }
    sstore(0x04, 0x0101)
}
```

In the above example, the nonZero case jumps to the `sstore(0x04, 0x0101)` block, whereas the `zero` 'directly' goes to the `store(0x04, 0x0101)` block. Shouldn't it be the other way around (based on how current codegen does it)?

I'm also okay with always having jumps in all cases, and deciding how to lay out code so that we can skip one of the jumps during the code transform.

See [visualizer](https://dreampuf.github.io/GraphvizOnline/#%20%20%20digraph%20CFG%20%7B%0A%20%20%20%20nodesep%3D0.7%3B%0A%20%20%20%20node%5Bshape%3Dbox%5D%3B%0A%0A%20%20%20%20Entry%20%5Blabel%3D%22Entry%22%5D%3B%0A%20%20%20%20Entry%20-%3E%20Block0%3B%0A%20%20%20%20Block0%20%5Blabel%3D%22%5C%0A%20%20%20%20sstore%3A%20%5B%200x00%200x00%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20sload%3A%20%5B%200x00%20%5D%20%3D%3E%20%5B%20TMP%5Bsload%2C%200%5D%20%5D%5Cl%5C%0A%20%20%20%20Assignment(GHOST%5B0%5D)%3A%20%5B%20TMP%5Bsload%2C%200%5D%20%5D%20%3D%3E%20%5B%20GHOST%5B0%5D%20%5D%5Cl%5C%0A%20%20%20%20eq%3A%20%5B%20GHOST%5B0%5D%200x00%20%5D%20%3D%3E%20%5B%20TMP%5Beq%2C%200%5D%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block0%20-%3E%20Block0Exit%3B%0A%20%20%20%20Block0Exit%20%5Blabel%3D%22%7B%20TMP%5Beq%2C%200%5D%7C%20%7B%20%3C0%3E%20Zero%20%7C%20%3C1%3E%20NonZero%20%7D%7D%22%20shape%3DMrecord%5D%3B%0A%20%20%20%20Block0Exit%3A0%20-%3E%20Block1%3B%0A%20%20%20%20Block0Exit%3A1%20-%3E%20Block2%3B%0A%0A%20%20%20%20Block1%20%5Blabel%3D%22%5C%0A%20%20%20%20sstore%3A%20%5B%200x0101%200x04%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block1Exit%20%5Blabel%3D%22MainExit%22%5D%3B%0A%20%20%20%20Block1%20-%3E%20Block1Exit%3B%0A%0A%20%20%20%20Block2%20%5Blabel%3D%22%5C%0A%20%20%20%20sstore%3A%20%5B%200x0101%200x01%20%5D%20%3D%3E%20%5B%20%5D%5Cl%5C%0A%20%20%20%20%22%5D%3B%0A%20%20%20%20Block2%20-%3E%20Block2Exit%20%5Barrowhead%3Dnone%5D%3B%0A%20%20%20%20Block2Exit%20%5Blabel%3D%22Jump%22%20shape%3Doval%5D%3B%0A%20%20%20%20Block2Exit%20-%3E%20Block1%3B%0A%0A%20%20%20%20%7D%0A)

ekpyron replied:
The yul optimizer should transform that switch to an ``if`` anyways, so I'd tend to say better to not introduce special cases in the graph or in the code transform - at least for now, if it turns out things like that would better be done here or during code generation, we can still do it later on - there's generally loads of optimization opportunities left in the entire process.

ekpyron replied:
That being said, the code generator even now won't actually generate that jump :-).
The logic there is to just keep generating through jumps, unless the block to be jumped to is already generated (so the non-zero branch here will be staged for later generation and the zero branch will immediately follow the pre-switch branch without jump).

ekpyron replied:
Although it may actually even be possible to avoid the logic for that in the code generation step, generate that jump and hope that the legacy codegen inliner will remove the jump, but I haven't checked that that works in all cases :-).

hrkrshnn replied:
> That being said, the code generator even now won't actually generate that jump

```
======= <stdin> (EVM) =======

Pretty printed source:
object "object" {
    code {
        sstore(0, 0)
        switch sload(0)
        case 0 { sstore(5, 5) }
        sstore(0x04, 0x0101)
    }
}


Binary representation:
6000600055600054600081146012576018565b60056005555b50610101600455

Text representation:
    /* "<stdin>":16:17   */
  0x00
    /* "<stdin>":13:14   */
  0x00
    /* "<stdin>":6:18   */
  sstore
    /* "<stdin>":36:37   */
  0x00
    /* "<stdin>":30:38   */
  sload
    /* "<stdin>":48:49   */
  0x00
    /* "<stdin>":43:77   */
  dup2
  eq
  tag_2
  jumpi
    /* "<stdin>":23:77   */
  jump(tag_1)
    /* "<stdin>":43:77   */
tag_2:
    /* "<stdin>":69:70   */
  0x05
    /* "<stdin>":66:67   */
  0x05
    /* "<stdin>":59:71   */
  sstore
    /* "<stdin>":23:77   */
tag_1:
  pop
    /* "<stdin>":95:101   */
  0x0101
    /* "<stdin>":89:93   */
  0x04
    /* "<stdin>":82:102   */
  sstore
```

The old codetransform doesn't jump in the nonZero case, whereas it jumps in the zero case. Currently the graph does the opposite, right?

ekpyron replied:
You'd get something like the following which is neither better nor worse:
```
    /* "<stdin>":16:17   */
  0x00
    /* "<stdin>":13:14   */
  0x00
    /* "<stdin>":6:18   */
  sstore
    /* "<stdin>":36:37   */
  0x00
    /* "<stdin>":30:38   */
  sload
    /* "<stdin>":48:49   */
  0x00
    /* "<stdin>":43:77   */
  dup2
  eq
  tag_2
  jumpi
    /* "<stdin>":23:77   */
tag_1:
  pop
    /* "<stdin>":95:101   */
  0x0101
    /* "<stdin>":89:93   */
  0x04
    /* "<stdin>":82:102   */
  sstore
  
  ...eventually "stop" ...
  
    /* "<stdin>":43:77   */
tag_2:
    /* "<stdin>":69:70   */
  0x05
    /* "<stdin>":66:67   */
  0x05
    /* "<stdin>":59:71   */
  sstore
    /* "<stdin>":23:77   */
  jump(tag_1)
```

hrkrshnn commented on libyul/backends/evm/ControlFlowGraphBuilder.cpp:
```diff
@@ -0,0 +1,469 @@
+	if (switchCase.value)
+	{
+		CFG::BasicBlock& caseBranch = m_graph.makeBlock();
+		makeConditionalJump(makeValueCompare(*switchCase.value), caseBranch, afterSwitch);
```
``` suggestion
		CFG::BasicBlock& tmpBlock = m_graph.makeBlock();
		makeConditionalJump(makeValueCompare(*switchCase.value), caseBranch, tmpBlock);
		m_currentBlock = &tmpBlock;
		jump(afterSwitch);
```

Maybe I wasn't clear before. I think there is a jump missing here.

hrkrshnn commented on libyul/backends/evm/ControlFlowGraph.h:
```diff
@@ -0,0 +1,210 @@
+	std::list<Scope::Function const*> functions;
+
+	/// Container for blocks for explicit ownership.
+	std::list<BasicBlock> blocks;
```
Random question: is this a list instead of a vector only because you don't have to worry about pointer invalidation?

ekpyron replied:
Yep, exactly. If I made it any container that has to move stuff, the pointers in the graph could be invalidated at any time with no real way to fix.
